### ISSUE TITLE
Handle mouse events on PlotController instead of PlotModel (and elements)

### ISSUE THREAD

**Issue by objorke** [oxyplot/oxyplot#111]
**Imported from [http://oxyplot.codeplex.com/workitem/10132](http://oxyplot.codeplex.com/workitem/10132)**

**[objo](http://www.codeplex.com/site/users/view/objo)** 2014-02-17 08:08
We can achieve the same functionality by using a <code>PlotController</code>

<a href="https://oxyplot.codeplex.com/workitem/9625" rel="nofollow">https://oxyplot.codeplex.com/workitem/9625</a>

<ul>
<li>Better separation of controller logic and model </li><li>Less code </li><li>Better performance of PlotController </li></ul>

**[objo](http://www.codeplex.com/site/users/view/objo)** 2014-03-19 22:57
It seems to be quite convenient to use the events. Or is that valid only for the simple demos?

**[objo](http://www.codeplex.com/site/users/view/objo)** 2014-03-21 09:02
An alternative could be to add an <code>EnableMouseEventsInModel</code> property on the
<code>PlotController</code>. The default value could be false. When the value is false, the mouse events should not be raised on the
<code>PlotModel</code> and <code>PlotElement</code>s.

**[tibel](http://www.codeplex.com/site/users/view/tibel)** 2014-04-12 08:43
Hmm, still don't like it. The PlotModel and its child elements should not know anything about a mouse, keyboard or touch. Also it is easy to create memory leaks by events on child elements of the PlotModel when your plot is more dynamic (e.g. adding and
 removing annotations).

Therefore I would prefer following:

<ul>
<li>mouse/keyboard events exist only on <code>PlotController</code> and the event will tell you what element of the plot was e.g. clicked
</li><li>similar the <code>AxisChanged</code> should be on <code>PlotModel</code> and tell you what axis changed
</li><li>Ultimately no child element of the plotmodel should expose any events

</li></ul>

With this approach I don't have problems with adding/removing elements from the plotmodel and my event handlers.

**Comment by tibel** [comment]
I like the suggested solution.

**Comment by malambika** [comment]
hello tiebl im new to this xamairn..how to handle mouse evnt in oxyplot using  plotcontroller?

**Comment by VisualMelon** [comment]
Would it be possible to discuss how existing usage of mouse events could be refitted? For example, if there is a plot-annotation that the user should be able click-and-drag, how do we imagine that would be achieved? Currently this is easily achieved with events, and can be done without the risk of memory leaks easily. A controller/module separation would require a way to register a per-component (or per-type-of-component) controller along with a way for this controller to 'take control' (e.g. when the component is clicked, or some other UI action outside of the library), and this doesn't seem to be provided. That is, there is no native provision for doing anything other than displaying the tracker when a left-mouse button is pressed: the controller does not have a composable extension point that lets us combine controllers for the left-mouse down gesture.

Without such a system, it'll be infeasible to produce portable and cross-platform interactive plot components.

For the record, I do use mouse events for work, so I have a vested interest in the existing functionality. I'd be happy to 'update' my existing code to a better system, but it would need to provide the same capabilities we enjoy at the moment.

**Comment by Doeharrrck** [comment]
It is definitely needed to keep some events on the model layer that allow the plot model or it's children to notify about changes to keep the current functionality. For instance, I implemented the functionality of the [Mouse Events - LineAnnotation](https://github.com/ylatuya/oxyplot/blob/master/Source/Examples/ExampleLibrary/Examples/MouseEventExamples.cs) example by implementing a manipulator to achieve the same functionality. But what if I want to display the current position of the cursor line OUTSIDE the plot. Another example is coupling axes among several plots like [here](https://github.com/ylatuya/oxyplot/blob/master/Source/Examples/WPF/WpfExamples/Examples/CoupledAxesDemo/MainWindow.xaml.cs).
I think there must be a separation between events that handle user input like mouse or keyboard events on the one hand, and events that notify what's going on inside the plot on the other hand. Only the first group of events can be easily replaced by manipulator instances.

**Comment by nzain** [comment]
Many `PlotModel.MouseX` are now obsolete and link to this ticket. That doesn't help me... what is the new alternative? I read that the controller should be used. How? Even worse: the examples of this repository suggest the same deprecated code. https://github.com/oxyplot/oxyplot/blob/02a21c5ce7b1cc531120660181226662fb5b948d/Source/Examples/ExampleLibrary/Examples/MouseEventExamples.cs#L33

So much code is broken now, events being only one of the non-trivial topics. We have a large code base on top of OxyPlot, turned into a nightmare of compiler errors and warnings.

**Comment by VisualMelon** [comment]
@nzain this is a good and important question, and I'm afraid I can't give you a good answer. The link to this is just as an indication of why the events are so annotated. Indeed, we need to provide better examples for this, and for many events there is presently no obvious alternative.

https://github.com/oxyplot/oxyplot/blob/develop/Source/Examples/ExampleLibrary/Examples/PlotControllerExamples.cs is the canonical plot controller example. Note that you can use hittesting to determine which plot elements have been clicked etc. and dispatch accordingly. The default controller implementation is also instructive: https://github.com/oxyplot/oxyplot/tree/develop/Source/OxyPlot/PlotController

Note that you can disable the warnings locally with `#pragma warning disable CS0618` (re-enable with `#pragma warning disable CS0618`) and the events are not going anywhere soon so I would not worry about continuing to use them for the time being.

If you have specifics use-case which are not well supported by the plot controller (or it's not obvious how they can be supported), please do open individual issues for then. This will help inform which examples we provide and motivate further development of the controller system if necessary.

**Comment by nzain** [comment]
Thanks for the link to the `PlotControllerExample`. We are actually using `PlotController` already, mixed with `PlotModel.MouseX` events in the very same objects. I don't remember why, but it looks like some functionality was only accessible through either events XOR controller.

Too much of my code is not working anymore. I can't even see some very basic `RectangleAnnotation` objects anymore. Before I can investigate (and eventually create issues) I'll need a large time slot. For now we have to stick to version 2.0. I'll monitor this issue for updates.

**Comment by VisualMelon** [comment]
> Too much of my code is not working anymore. I can't even see some very basic RectangleAnnotation objects anymore. Before I can investigate (and eventually create issues) I'll need a large time slot. For now we have to stick to version 2.0. I'll monitor this issue for updates.

No problem. If you think there are any regressions in 2.1, or anything that should be mentioned in the release notes but isn't, please do report them as well as separate issues or ping me on the [gitter](https://gitter.im/oxyplot/oxyplot).

**Comment by trockefeller-pathway** [comment]
I think I have figured out how to replace the event handlers for MouseDown with something like below

```csharp
Controller.BindMouseDown(OxyMouseButton.Left, new DelegateViewCommand<OxyMouseDownEventArgs>(myPlot_HandleMouseDown));

private void myPlot_HandleMouseDown(IView view, IController controller, OxyMouseDownEventArgs e)
{
    //handle mouse down stuff
}
```

But when trying to do something similar for `MouseUp` and `MouseMove`, I am finding that there is no `BindMouseUp` or `BindMouseMove`, but instead there are methods `HandleMouseUp` and `HandleMouseMove`

How do I replace those event handlers that were originally written something like this

```csharp
myPlotModel.MouseUp += myPlotModel_MouseUp;
myPlotModel.MouseMove += myPlotModel_MouseMove;
```

**Comment by VisualMelon** [comment]
@trockefeller-pathway I don't think you can: the idea is that MoustDown, MouseMove, and MouseUp are all handled by a single manipulator with methods Started, Delta, and Completed.

**Comment by VisualMelon** [comment]
@trockefeller-pathway yeah, that's the idea. See https://github.com/oxyplot/oxyplot/blob/develop/Source/OxyPlot/PlotController/Manipulators/PanManipulator.cs for a full example of a mouse manipulator implementation.

**Comment by Jkallus** [comment]
@trockefeller-pathway Did you find a path forward?

**Comment by TimurVa** [comment]
@VisualMelon  It's still unclear for me. How we should handle mouse down/move/up on Line annotation for example with AddMouseManipulator as all events are deprecated?

**Comment by Spordoz** [comment]
@TimurVa You can handle them in the manipulator as displayed above and do a hit test on your Annotation during the respective function (started for down, delta for move, completed for up)

**Comment by carlreinke** [comment]
Is the intent to remove `Axis.AxisChanged` or to move it?  The `Obsolete` message says "remove," but the [issue](https://github.com/oxyplot/oxyplot/issues/111#issue-40306938) says it should be moved to `PlotModel`.

**Comment by patrick-stalph** [comment]
Four years later and we still rely on OxyPlot, because it is just great. By the time I've learned to override `PlotController.HandleMouseDown(..)` and related events to achieve the same functionality instead of using any of the deprecated events. While I understand the reason for this planned change, the code is clunky and harder to understand/maintain. It becomes obvious, when you look at the mouse event and plot controller samples, which *all* use the deprecated events. I would not dare to submit a PR with rewritten samples, because it gets ugly even for such simple single-topic samples.
